### PR TITLE
Fix reproducible linux builds 

### PR DESCRIPTION
**Pull request by niccoloraspa**
Closes: https://github.com/osmosis-labs/osmosis/issues/2362

## What is the purpose of the change

This PR:

- Improves the main Dockerfile 
- Fixes reproducible builds in the Makefile

**🐳 General Dockerfile improvements:**

- Move up the layer to download `go` dependencies
- Automatically detect CosmWasm version and architecture to download the correct `libwasm`
- Explicitly use the `go build...` command
- Generalize runner images

**🛠 Reproducible build** idea:

- Compile the binary (statically linked with CosmWasm) while building the container image
- `cp` the binary out of the container and place it in the build folder

These simple changes drastically simplify the build process, and they allow very nice improvements:

1. `osmobuilder` is now deprecated as the functionality is now integrated in the main `Dockerfile`

> Removing `osmobuilder` would requires:
> - [ ] Changing the state-compatibility-check to build the binary using `make build-reproducible-amd64`
> - [ ] Update release docs

2. Deprecating `osmobuilder` allows to deprecated the whole `contrib/` folder, that can now be removed. 

> We would need to move the `scripts` 

4. We can now backport this build process to all `vX` branches to have consistent way to build the `osmosisd` binary across all versions (only difference would be the CosmWasm line)

5. The generalization of the runner image allows to create different flavour of the images starting from the same Dockerfile. This can be used to:
- Create different flavours of the `osmosis` Docker image (`distroless`, `nonroot`, `alpine`)
- Reduce the number of Dockerfile used internally (ref: https://github.com/osmosis-labs/osmosis/issues/1679)

I will create issues for every point.

## Brief Changelog

- Improve `Dockerfile`
- Fix commands in `Makefile` to create reproducible builds for linux

## Testing and Verifying

You can build the binaries with:

```
make build-reproducible
# Inspected build folder 
ls build
# osmosisd-linux-amd64 osmosisd-linux-arm64
```

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

**Comment by czarcas7ic** [comment]
Looks like we will need to modify e2e tests for this change, will look into it

**Comment by niccoloraspa** [comment]
Thanks @czarcas7ic

**Review by czarcas7ic — APPROVED** [review]
Seems like its good to go now, just had to specify start command in e2e now that we took it out form the Dockerfile (I think that was the right move FWIW)

File: Dockerfile [hunk 1]
```diff
@@ -1,50 +1,61 @@
...
-  sha256sum /lib/libwasmvm_muslc.x86_64.a | grep f6282df732a13dec836cda1f399dd874b1e3163504dbd9607c6af915b2740479
-# Copy the right library according to architecture. The final location will be found by the linker flag `-lwasmvm_muslc`
-RUN cp /lib/libwasmvm_muslc.$(uname -m).a /lib/libwasmvm_muslc.a
-
-# Get go dependencies first
-COPY go.mod go.sum /osmosis/
+# Download go dependencies
 WORKDIR /osmosis
+COPY go.mod go.mod
+COPY go.sum go.sum
```
Comment (p0mvn) [inline Dockerfile:17]: Consider doing this in one command to minimize the number of layers. As per:
https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

**Review by p0mvn — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (niccoloraspa) [reply Dockerfile:17]: Thanks! I'add in the next PR I'm making so we don't have to review again the changes

File: Dockerfile [hunk 2]
```diff
@@ -1,50 +1,61 @@
...
+      -tags "netgo,ledger,muslc" \
+      -ldflags "-X github.com/cosmos/cosmos-sdk/version.Name="osmosis" \
+              -X github.com/cosmos/cosmos-sdk/version.AppName="osmosisd" \
+              -X github.com/cosmos/cosmos-sdk/version.Version=$VERSION \
+              -X github.com/cosmos/cosmos-sdk/version.Commit=$COMMIT \
+              -X github.com/cosmos/cosmos-sdk/version.BuildTags='netgo,ledger,muslc' \
+              -w -s -linkmode=external -extldflags '-Wl,-z,muldefs -static'" \
+      -trimpath \
+      -o /osmosis/build/ \
+      ./...
```
Comment (ValarDragon) [inline Dockerfile:50]: Is there a make command we should be running, instead of copying this here?

Comment (niccoloraspa) [reply Dockerfile:50]: Not sure if there is already a `make` target that will generate the `go build` command with all these flags.

We could for sure write one but avoiding using the Makefile in the Dockerfile was intentional. It makes the image creation more transparent and immediate.

**Pull request merged**